### PR TITLE
add install_method parameter

### LINKED ISSUE #103: Unrecognized keyword: @preserve with ubuntu1604
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: Any
- Ruby:
- Distribution: Ubuntu1604
- Module version:

## How to reproduce (e.g Puppet code you use)

use `preserve_chains_in_tables`

## What are you seeing

Impossible to use `preserve_chains_in_tables` , what logs display:
```
 systemd[1]: Stopping ferm firewall configuration...
 ferm[7903]:  * Stopping Firewall ferm
 ferm[7903]: Error in /etc/ferm/ferm.conf line 9:
 ferm[7903]:     ip ip6
 ferm[7903]: )
 ferm[7903]: table nat
 ferm[7903]: {
 ferm[7903]:     chain DOCKER @preserve <--
 ferm[7903]: Unrecognized keyword: @preserve
 systemd[1]: ferm.service: Control process exited, code=exited status=255
 systemd[1]: Stopped ferm firewall configuration.
 systemd[1]: ferm.service: Unit entered failed state.
 systemd[1]: ferm.service: Failed with result 'exit-code'.
```

## Any additional information you'd like to impart

The `ferm` version shipped with Ubuntu1604 is 2.2 and does not know `@preserve`.

### PR DESCRIPTION
**Pull request by Dan33l**
#### Pull Request (PR) description

This PR add parameter `install_method` that permits to choose between `package` and `vcsrepo`. This permit to install more recent version of `ferm` when desired.

#### This Pull Request (PR) fixes the following issues

Fixes #103

### CODE REVIEW

File: manifests/config.pp [hunk 1]
```diff
@@ -10,6 +10,25 @@
 
   $_ip = join($ferm::ip_versions, ' ')
 
+  if $ferm::install_method == 'vcsrepo' and $ferm::manage_service {
+    include systemd::systemctl::daemon_reload
+
+    file_line { "ferm.service use ${ferm::configfile} to start":
```
Comment (bastelfreak) [inline manifests/config.pp:16]: instead of modyfiying the default unit file, we should always replace it with a version in /etc, or for this specific usecase, use a drop_in: https://github.com/camptocamp/puppet-systemd#drop-in-files

**Comment by Dan33l** [comment]
Note to myself, the CentOS6 is failing to start service `ferm` with message :
```
starting firewall fermFATAL: Could not load /lib/modules/4.15.0-74-generic/modules.dep: No such file or directory
FATAL: Could not load /lib/modules/4.15.0-74-generic/modules.dep: No such file or directory
FATAL: Could not load /lib/modules/4.15.0-74-generic/modules.dep: No such file or directory
FATAL: Could not load /lib/modules/4.15.0-74-generic/modules.dep: No such file or directory
```
The directory is empty.

File: manifests/config.pp (hunk 1 above)
Comment (Dan33l) [reply manifests/config.pp]: can you confirm this is now addressed ?

**Comment by Dan33l** [comment]
On `debian10`, it looks that `iptables-save` point to `/usr/sbin/xtables-nft-multi`.

The acceptance test uses `iptables-save` to verify rules. But to check rules with `install_method => vcsrepo` we have to use `iptables-legacy-save` instead of `iptables-save`

File: manifests/install.pp [hunk 2]
```diff
@@ -8,8 +8,45 @@
...
+      }
+    }
+    'vcsrepo': {
+      $_source_path = '/opt/ferm'
+
+      package{'ferm':
+        ensure => absent,
+      }
+
+      -> package { ['iptables', 'perl', 'make']:
```
Comment (bastelfreak) [inline manifests/install.pp:24]: at least perl and make could be managed by another module, should be use ensure_package here?

Comment (Dan33l) [reply manifests/install.pp]: list of managed packages is now an `Array` as parameter

File: spec/acceptance/ferm_spec.rb [hunk 3]
```diff
@@ -101,7 +135,7 @@ class { 'ferm':
           require           => Ferm::Chain['check-http'],
         }
       )
-      pp = [basic_manifest, advanced_manifest].join("\n")
+      pp = [basic_manifest, '}', advanced_manifest].join("\n")
```
Comment (dhoppe) [inline spec/acceptance/ferm_spec.rb:138]: Is this supposed to look like this.

File: manifests/install.pp (hunk 2 above)
Comment (Dan33l) [reply manifests/install.pp]: The function [ensure_packages()](https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/REFERENCE.md#ensure_packages) is now used.

File: manifests/config.pp [hunk 4]
```diff
@@ -10,6 +10,19 @@
 
   $_ip = join($ferm::ip_versions, ' ')
 
+  if $facts['systemd'] { #fact provided by systemd module
+    if $ferm::install_method == 'vcsrepo' and $ferm::manage_service {
+      include systemd::systemctl::daemon_reload
```
Comment (bastelfreak) [inline manifests/config.pp:15]: this isn't required. systemd::dropin_file does the notify:
https://github.com/camptocamp/puppet-systemd/blob/master/manifests/dropin_file.pp#L93-L103

File: manifests/init.pp [hunk 5]
```diff
@@ -67,6 +70,9 @@
   Hash $chains = {},
   Array[Enum['ip','ip6']] $ip_versions = ['ip','ip6'],
   Hash[String[1],Array[String[1]]] $preserve_chains_in_tables = {},
+  Enum['package','vcsrepo'] $install_method = 'package',
+  String $vcsrepo = 'https://github.com/MaxKellermann/ferm.git',
+  String $vcstag = 'v2.5.1',
```
Comment (bastelfreak) [inline manifests/init.pp:75]:
```suggestion
  String[1] $vcstag = 'v2.5.1',
```

File: manifests/init.pp [hunk 6]
```diff
@@ -67,6 +70,9 @@
   Hash $chains = {},
   Array[Enum['ip','ip6']] $ip_versions = ['ip','ip6'],
   Hash[String[1],Array[String[1]]] $preserve_chains_in_tables = {},
+  Enum['package','vcsrepo'] $install_method = 'package',
+  String $vcsrepo = 'https://github.com/MaxKellermann/ferm.git',
```
Comment (bastelfreak) [inline manifests/init.pp:74]:
```suggestion
  Stdlib::HTTPSUrl $vcsrepo = 'https://github.com/MaxKellermann/ferm.git',
```

File: manifests/install.pp [hunk 7]
```diff
@@ -8,8 +8,40 @@
...
-    ensure => 'latest',
+  case $ferm::install_method {
+    'package': {
+      package{'ferm':
+        ensure => 'latest',
+      }
+    }
+    'vcsrepo': {
+      $_source_path = '/opt/ferm'
+      ensure_packages (['iptables', 'perl', 'make'], { ensure => present })
```
Comment (bastelfreak) [inline manifests/install.pp:19]: isn't `git` also a requirement that should be installed?

File: spec/acceptance/ferm_spec.rb [hunk 8]
```diff
@@ -43,12 +53,36 @@ class { 'ferm':
       },
     },
     ip_versions      => ['ip'], #only ipv4 available with CI
-  }
```
Comment (bastelfreak) [inline spec/acceptance/ferm_spec.rb:46]: why did you delete this line?

File: templates/dropin_ferm.conf.epp [hunk 9]
```diff
@@ -0,0 +1,5 @@
+[Service]
```
Comment (bastelfreak) [inline templates/dropin_ferm.conf.epp:1]: please add a comment like `# THIS SNIPPET IS MANAGED BY PUPPET`

File: spec/acceptance/ferm_spec.rb (hunk 8 above)
Comment (Dan33l) [reply spec/acceptance/ferm_spec.rb:46]: to factorize the manifest and add at this end some parameters or not

File: spec/spec_helper_acceptance.rb [hunk 10]
```diff
@@ -2,4 +2,5 @@
 
 configure_beaker do |host|
   install_package(host, 'epel-release') if fact_on(host, 'os.name') == 'CentOS'
+  install_package(host, 'git')
```
Comment (bastelfreak) [inline spec/spec_helper_acceptance.rb:5]: this isn't required anymore since we now manage the git package.

**Comment by Dan33l** [comment]
@bastelfreak thank you for the review
